ethereum/contracts/query/QueryDemo.sol

```diff
@@ -0,0 +1,103 @@
+// contracts/query/QueryDemo.sol
+// SPDX-License-Identifier: Apache 2
+
+pragma solidity ^0.8.0;
+
+import "../libraries/external/BytesLib.sol";
+import "../interfaces/IWormhole.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/Context.sol";
+import "./QueryResponse.sol";
+
+/// @dev QueryDemo is a library that implements the parsing and verification of Cross Chain Query (CCQ) responses.
+contract QueryDemo is Context, QueryResponse {
+    using BytesLib for bytes;
+
+    struct ChainEntry {
+        uint16 chainID;
+        address contractAddress;
+        uint256 counter;
+        uint256 blockNum;
+        uint256 blockTime;
+    }
+
+    address owner;
+    address wormhole;
+    uint16 myChainID;
+    mapping(uint16 => ChainEntry) counters;
+    uint16[] foreignChainIDs;
+
+    bytes4 GetMyCounter = bytes4(hex"916d5743");
+
+    constructor(address _owner, address _wormhole, uint16 _myChainID) {
+        owner = _owner;
+        wormhole = _wormhole;  
+        myChainID = _myChainID;
+        counters[myChainID] = ChainEntry(myChainID, address(this), 0, 0, 0);
+    }
+
+    // updateRegistration should be used to add the other chains and to set / update contract addresses.
+    function updateRegistration(uint16 _chainID, address _contractAddress) public onlyOwner {
+        if (counters[_chainID].chainID == 0) {
+            foreignChainIDs.push(_chainID);
+            counters[_chainID].chainID = _chainID;
+        }
+
+        counters[_chainID].contractAddress = _contractAddress;
+    }
+
+    // getMyCounter (call signature 916d5743) returns the counter value for this chain. It is meant to be used in a cross chain query.
+    function getMyCounter() public view returns (uint256) {
+        return counters[myChainID].counter;
+    }
+
+    // getState() returns this chain's view of all the counters. It is meant to be used in the front end.
+    function getState() public view returns (ChainEntry[] memory) {
+        ChainEntry[] memory ret = new ChainEntry[](foreignChainIDs.length + 1);
+        ret[0] = counters[myChainID];
+        for (uint idx=0; idx<foreignChainIDs.length; idx++) {
+            ret[idx+1] = counters[foreignChainIDs[idx]];
+        }      
+
+        return ret;
+    }
+
+    // updateCounters takes the cross chain query response for the two other counters, stores the results for the other chains, and updates the counter for this chain.
+    function updateCounters(bytes memory response, IWormhole.Signature[] memory signatures) public {
+        uint256 adjustedBlockTime;
+        ParsedQueryResponse memory r = parseAndVerifyQueryResponse(address(wormhole), response, signatures);
+        require(r.responses.length == foreignChainIDs.length, "unexpected number of results");
+        for (uint idx=0; idx<r.responses.length; idx++) {
```

Comment:
Is it possible for there to be enough records in r.responses for this to consume the block gas limit? Should there be a strict upper bounds on the number of responses? With so many storage reads and updates, this seems like it could be an expensive function.

Reply:
The number of responses is limited to 255, and enforced by the guardian. If that turns out to require too much gas, we could enforce something here, but for this demo, I don't think that should be an issue.

Reply:
Ah totally thanks. I just see a lot of DoS / Gas griefing and always am leery of unbounded loops in solidity contracts with user supplied data.

bounding at 255 max would be way below the gas limits, thanks.